lib/maths/CBoostedTreeFactory.cc

```diff
@@ -306,14 +325,19 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) {
         m_TreeImpl->m_MaximumNumberTrees = *(m_TreeImpl->m_MaximumNumberTreesOverride);
     } else {
         // This needs to be tied to the learn rate to avoid bias.
-        m_TreeImpl->m_MaximumNumberTrees =
-            static_cast<std::size_t>(2.0 / m_TreeImpl->m_Eta + 0.5);
+        m_TreeImpl->m_MaximumNumberTrees = computeMaximumNumberTrees(m_TreeImpl->m_Eta);
     }
 
     if (m_TreeImpl->m_FeatureBagFractionOverride != boost::none) {
         m_TreeImpl->m_FeatureBagFraction = *(m_TreeImpl->m_FeatureBagFractionOverride);
     }
 
+    double numberFeatures{static_cast<double>(m_TreeImpl->m_Encoder->numberEncodedColumns())};
+    double downSampleFactor{200.0 * numberFeatures /
```

Comment:
Ok so this is derived from the usual rule of thumb that the number of examples for linear regression should be proportional to the number of regressors to ensure independence of *empirical* residuals. I would normally say I need 30 * number of regressors to be confident this condition holds, so this represents a large comfort margin. However, since 1) we have a more expressive model, 2) we use this to upper bound the range of downsampling factors we search using a higher value seems appropriate. On the empirical side, for the data sets I tested we didn't choose values higher than this in the line search (again with some margin), but this could do with more testing. Perhaps the best thing to do would be to expose this parameter in the API and we should also do a fuller study down-the-line.

Reply:
I think 200 is a good number. Funny, I've heard about the rule of thumb with 30x in the context of neural networks years ago. Could never confirm where it originated from :smile: Could you please extract it as a constant with a meaningful name? I wouldn't mind to keep it hidden from users, as it is very technical, but we would need to monitor it on real data.

Reply:
I kept the default but exposed the parameter as a get out-of-jail of this is very wrong. See [5b9a04a](https://github.com/elastic/ml-cpp/pull/811/commits/5b9a04aafe6d1c30ce7ad3b31df23b2cd82a92f8).